### PR TITLE
DATA-876 refactor inverse kinematics to become usable outside of motionplan package

### PR DESCRIPTION
**Pull request by raybjork**
This PR adds utility functionality to the inverse kinematics objects in `motionplan` to allow it to be usable outside of motionplan. Specifically this is a prerequisite to upgrade the version of this repo used in the [motion-testing repo](https://github.com/viamrobotics/motion-testing).  

To accomplish this, the major changes involve:
- refactoring IK structures to be more modular and usable
- adding a public function `BestIKSolutions` to expose our IK solver to users outside the `motionplan` package
- reworking of the planner options to differentiate between ik parameters and planning parameters
- `plannerManager` rework to allow `BestIKSolutions` to tie into the same logic and avoid repeating ourselves
- add method to `Frame` interface to make the distinction between `solverFrame` and other `Frames` negligible

### CODE REVIEW

File: motionplan/plannerOptions.go
Comment (biotinker) [inline motionplan/plannerOptions.go:102]: We should not be making this public; allowing users to hand-construct these structs is a disaster waiting to happen.

Instead, the public constructors for IK methods should take the same map[string]interface{} that motion.Solve does, and resolve those to the desired options the same way.

File: examples/test/main.go
Comment (raybjork) [inline examples/test/main.go:29]: Oops I didn't mean to add the files in the `examples` folder.  Will remove

File: motionplan/plannerOptions.go
Comment (raybjork) [reply motionplan/plannerOptions.go:102]: Yeah I didn't like doing this either and did this as a temporary workaround so I could get feedback on the rest of the project. I like your idea better

File: examples/test/main.go
Comment (tessavitabile) [reply examples/test/main.go:29]: I also recommend changing secrets on the pi, since this PR is public.

**Review by tessavitabile — COMMENTED** [review]
For my own understanding, what's the relationship between `InverseKinematicsSolver`, `Motion.Move()`, CBiRRT and RRT*? Is it something like: `Motion.Move()` constructs an `InverseKinematicsSolver` which creates plans using either CBiRRT or RRT* depending on its configuration.

**Comment by biotinker** [comment]
@tessavitabile 

> what's the relationship between InverseKinematicsSolver, Motion.Move(), CBiRRT and RRT*?

Your guess is close. `Motion.Move()` constructs an `InverseKinematicsSolver`, but that does not in turn create the motion planner (RRT* and CBiRRT) but rather exists in parallel to them. I would break it apart like so:

- `Motion.Move()` receives a set of joint positions as a starting state, and a cartesian position as a goal state.
- The `InverseKinematicsSolver` takes that cartesian goal state, and creates N valid joint states that will yield it
- The motion planners take the start state, the list of goal states, and are responsible for connecting the two without bumping into obstacles
- The Planager is what manages the above two bullet points.

**Comment by tessavitabile** [comment]
Thanks, @biotinker, that makes sense!

File: examples/test/main.go
Comment (raybjork) [reply examples/test/main.go]: ...oops. Good catch

File: motionplan/plannerOptions.go
Comment (raybjork) [reply motionplan/plannerOptions.go]: With the refactor, this comment should be addressed.  Going to resolve

File: motionplan/inverseKinematics.go
Comment (biotinker) [inline motionplan/inverseKinematics.go:18]: This should be private

File: motionplan/inverseKinematics.go
Comment (biotinker) [inline motionplan/inverseKinematics.go:54]: If NumThreads == 0, we should use the default CPU count, not 1.

File: motionplan/inverseKinematics.go
Comment (biotinker) [inline motionplan/inverseKinematics.go:66]: randseed should either use a default, or be taken from IK options

File: motionplan/inverseKinematics.go
Comment (biotinker) [inline motionplan/inverseKinematics.go:38]: This is a weird data flow.

The user is expected to manually create a Solver using this function, but can't actually use it (all member methods are private). They then have to pass this into BestIKSolutions.

Rather than having this two-step process, we should just fold this into the `BestIKSolutions` call.

File: motionplan/rrt.go
Comment (biotinker) [inline motionplan/rrt.go:23]: I'm a little wary of doing this abstraction.

We're definitely going to eventually have non-RRT planners, so distinguishing RRT and non-RRT planners makes sense.

But it's entirely possible we'll never have non-parallel RRT planners. In which case we're just adding an extra layer of abstraction for no reason.

File: motionplan/inverseKinematics.go
Comment (raybjork) [reply motionplan/inverseKinematics.go:37]: Yeah I like this better too.  Will change

File: motionplan/rrt.go
Comment (raybjork) [reply motionplan/rrt.go:23]: I agree.  I meant to call this `parallelMotionPlanner` and have it only depend on `motionPlanner` about to push up some code illustrating this

File: motionplan/ensembleIKSolver.go
Comment (spieswl) [inline motionplan/ensembleIKSolver.go:32]: Related to Peter's suggestion, you could specify the seed for each new NLOptIKSolver here if you had a "starting seed" as part of the `opts` struct.

Comment (biotinker) [reply motionplan/ensembleIKSolver.go:32]: It's tough to build it into the solver struct itself, because the solver struct ought generally be thread-safe, while rand seeds are not

File: motionplan/inverseKinematics.go
Comment (biotinker) [reply motionplan/inverseKinematics.go:127]: I think my preferred way to do this is: put `randseed` in the `ikConfig` generic map input, and use a default of 1 if the user doesn't pass anything in. Since IK behavior is defined not only by the random seed but also by the seed input, this should work fine.

Motion planning should call `getSolutions` anyway rather than this.

File: motionplan/inverseKinematics.go
Comment (biotinker) [inline motionplan/inverseKinematics.go:138]: We can't assume this. We need to construct the frame system using the additional transforms in WorldState, place obstacles accordingly based on that. Specifying the parentage of the frame we're solving for might be needed as well.

File: motionplan/inverseKinematics.go
Comment (spieswl) [inline motionplan/inverseKinematics.go:46]: A potential usability improvement would be passing in the `ikConfig` map as an argument and handling that before you initially return `opts. 

There's precedent for doing this AND adding the JSON marshal/unmarshal code, such as what is done with cbirrt options. Compare L97-104 in this file versus L55-62 of cBiRRT.go in the newCbirrtOptions method). Not critical, but might save repeating a few lines in newIKSolver.

File: motionplan/inverseKinematics.go
Comment (biotinker) [reply motionplan/inverseKinematics.go:138]: For example, in the builtin motion service, the thing being solved for can be parented to things defined in additional transforms.

File: motionplan/nloptIKSolver.go
Comment (spieswl) [inline motionplan/nloptIKSolver.go:99]: I'm not seeing any place where we change the `id`, despite having logic for id == 1 later on. Is this still necessary?

Comment (biotinker) [reply motionplan/nloptIKSolver.go:99]: That's inside the ensembleIKsolver

File: motionplan/ensembleIKSolver.go
Comment (spieswl) [reply motionplan/ensembleIKSolver.go:32]: OK, fair enough

File: motionplan/nloptIKSolver.go
Comment (spieswl) [reply motionplan/nloptIKSolver.go:99]: Yep, found it, thanks for the pointer.

File: motionplan/inverseKinematics.go
Comment (raybjork) [reply motionplan/inverseKinematics.go:44]: I think this is mostly taken care of in the latest update

File: motionplan/inverseKinematics.go
Comment (raybjork) [reply motionplan/inverseKinematics.go]: The changes I've just introduced address this

**Comment by tessavitabile** [comment]
Removing myself from reviewers, as I don't plan to do a thorough review.

**Comment by tessavitabile** [comment]
Github is not letting me remove myself :( feel free to merge without my approval

File: motionplan/ensembleIKSolver.go
Comment (biotinker) [inline motionplan/ensembleIKSolver.go:75]: Why split this line?

File: motionplan/inverseKinematics.go
Comment (biotinker) [inline motionplan/inverseKinematics.go:200]: `options` should be cached outside the for loop, rather than making a function call each iteration

File: motionplan/motionPlanner.go
Comment (biotinker) [inline motionplan/motionPlanner.go:31]: We shouldn't be pre-emptively doing this, as we don't know what parallel non-RRT algorithms will look like. Let's stick with `motionPlanner` and `rrtParallelPlanner` rather than `rrtMotionPlanner` and `parallelMotionPlanner`. We can reorganize those if the need comes up.

File: motionplan/inverseKinematics.go
Comment (biotinker) [inline motionplan/inverseKinematics.go:41]: This should not be in `ikOptions`, but rather in a more generic struct that `ikOptions` inherits from, along with `extra` and `constraintHandler`. Otherwise, it will be impossible for planners to use these things without creating an `ikOptions`, regardless of whether the planner does IK.

File: referenceframe/frame.go
Comment (biotinker) [inline referenceframe/frame.go:125]: We've already got the public function `GetFrameInputs` that does exactly this.

We shouldn't be putting methods on interfaces to do super trivial convenience functions like this. Every time we do, it makes it harder to implement anything that doesn't inherit from the private base struct.

**Review by biotinker — CHANGES_REQUESTED** [review]
Found a few things that should be rearranged/updated.

Also, please generate quality data on both current `main` and on this branch and post the resulting charts; we want to make sure there are no regressions we aren't seeing, and the updates here are only needed to run OMPL in the planner quality repo, not RDK planners.

File: motionplan/inverseKinematics.go
Comment (raybjork) [reply motionplan/inverseKinematics.go:200]: changed

File: motionplan/motionPlanner.go
Comment (raybjork) [reply motionplan/motionPlanner.go:31]: I'm good to rename back to rrtParallelPlanner, but I don't want to reorganize back to just `rrtMotionPlanner`.  As it currently stands I've abstracted out all the rrt specific code to the rrt.go function and I would like to keep it this way as it makes it easier to build on, regardless of if we have another type of algorithm

File: referenceframe/frame.go
Comment (raybjork) [reply referenceframe/frame.go:125]: Currently we have to use switch statements to disambiguate the solver frame from other frames.  Its way uglier to have to call out the case (in a few places of the codebase at this point) where a frame might not be supported by GetFrameInputs, than the small overhead we incur if we happen to write a frame that doesn't inherit from `baseFrame`

Comment (raybjork) [reply referenceframe/frame.go:125]: This change rectifies things so that you can pass around the more general `Frame`, and not have to worry about what type of frame it is

Comment (raybjork) [reply referenceframe/frame.go:125]: Thinking about this more, there could be a middle ground in which we keep GetFrameInputs and do the case checking there

File: motionplan/inverseKinematics.go
Comment (raybjork) [reply motionplan/inverseKinematics.go:41]: In the same way you don't want to prematurely optimize for non-RRT planners, I would rather not make this change and cater to imagined future planners that don't use IK until we actually have an instance of this.

File: referenceframe/frame.go
Comment (raybjork) [reply referenceframe/frame.go:125]: Looked into the above option and it's also a little weird.  We would like to be able to call this from `referenceframe`, but since `solverFrames` exist outside this package we can't do the type check.  I'm back to thinking that adding the method is the cleanest solution

Comment (biotinker) [reply referenceframe/frame.go:125]: The only reason this change is needed, is that the planager was updated to take a generic frame rather than a solver frame.

Since we have tightly controlled entry into all planning now via the planager, there's no need for this. All input frames are wrapped in solver frames.

We should undo the generic frame change, swap all planning back to operating on solverFrames (wrapping frames with solver frames if need be), and then we don't need this at all.

File: motionplan/inverseKinematics.go
Comment (spieswl) [inline motionplan/inverseKinematics.go:25]: Nit: In some places we set defaultNumThreads as NumCPU / 2, others we set it as NumCPU / 4 (with some caveats in places). We should settle on one calculation for this.

File: motionplan/inverseKinematics.go
Comment (spieswl) [reply motionplan/inverseKinematics.go:41]: Does this even need to have `ik` in the name? What are other planners going to produce that aren't already captured in metrics, scores, and solutions?

File: motionplan/inverseKinematics.go
Comment (raybjork) [reply motionplan/inverseKinematics.go:25]: `Looks like its one other place, in some test code.  I'll change it to be `NumCPU/2`

File: motionplan/solverFrame.go
Comment (spieswl) [inline motionplan/solverFrame.go:249]: Nit: Docs mismatch.

File: motionplan/solverFrame.go
Comment (spieswl) [inline motionplan/solverFrame.go:263]: It seems strange to me to have a private `inputToMap` and a public `InputFromMap` which both work on solverFrame pointers. I haven't read the big thread in `frame.go` yet, but I think some users would be annoyed to see it won't go both ways.

File: referenceframe/frame.go
Comment (spieswl) [reply referenceframe/frame.go:125]: I started writing a long thing about solverFrames and the extent of their lifetimes, yadda, yadda.

It doesn't matter. Peter is right. It's far more sustainable to constrain `solverFrame` and all its structural nuance to the `motionplan` package and do sanitization of input frames so the user doesn't have to start keeping multiple frame styles within their mind.

**Review by spieswl — CHANGES_REQUESTED** [review]
Just adding some minor consistency improvements that we could stand to make.

File: motionplan/inverseKinematics.go
Comment (biotinker) [reply motionplan/inverseKinematics.go:41]: It's fine not to cater to nonexistent planners, but we should organize the fields such that they make sense where they are used; it doesn't make sense to have a motion planner look in its IK options for the number of CPU threads to use.

Comment (biotinker) [reply motionplan/inverseKinematics.go:41]: `MinScore` and `MaxSolutions` are the only IK specific things, so if we want IK specific options, those are what should be in it.

The other fields should migrate back to base `plannerOptions`, which IKoptions should inherit from, not vice versa.

File: motionplan/solverFrame.go
Comment (raybjork) [reply motionplan/solverFrame.go:263]: Yeah, me and Peter have been discussing a bit what the best approach is for this so I'm going to wait on the outcome of that before making this change.

File: referenceframe/primitives.go
Comment (raybjork) [inline referenceframe/primitives.go:77]: @biotinker Just realized this function will error if we pass a `solverFrame` which although is valid input seems like another point in favor of "get rid of this and tie it to the frame its operating on"

**Comment by raybjork** [comment]
@biotinker @tessavitabile Per our discussion at the board grooming should I close this for the time being?

**Comment by tessavitabile** [comment]
Oh, I thought the decision was to finish it! If it's close to being ready and generally a good tech debt improvement. We put the ticket into the current sprint.

**Comment by biotinker** [comment]
@tessavitabile I also thought the decision was to close it for the time being? I thought Fahmina had said to hold off until a customer requests this.

This PR has a bunch of refactor way beyond just exposing IK as public that we may want to tech spec out a bit more- it's not ready to merge as-is

**Comment by tessavitabile** [comment]
Thanks, I misunderstood then. Closing makes sense.

**Comment by raybjork** [comment]
Per the above conversation, closing this for the time being, will reopen at a later time when we are ready to make these changes

**Pull request closed without merging**
